ambari-server/src/main/java/org/apache/ambari/server/topology/ClusterTopologyImpl.java

```
@@ -111,11 +113,42 @@ public ClusterTopologyImpl(
     stack = request.getStack();
     setting = request.getSetting();
     blueprint.getConfiguration().setParentConfiguration(stack.getConfiguration(getServices()));
+
+    checkForDuplicateHosts(request.getHostGroupInfo());
     registerHostGroupInfo(request.getHostGroupInfo());
   }
 
+  ClusterTopologyImpl copy() {
+    try {
+      return new ClusterTopologyImpl(ambariContext, provisionRequest, resolvedComponents);
+    } catch (InvalidTopologyException e) {
+      // cannot happen, since already validated
+      LOG.error("Failed to copy cluster topology {}", this);
+      return null;
+    }
```

Throwing an IllegalStateException instead of returning null?

---

Actually this method was not used, so I've removed it.